app/support/order_row_importer.rb

```diff
@@ -0,0 +1,206 @@
+  end
+
+  def self.headers_to_s
+    HEADERS.values.join(",")
+  end
+
+  def initialize(row, order_import)
+    @row = row
+    @order_import = order_import
+    @errors = Set.new
+  end
+
+  def account
+    @account ||=
+      user
+        .accounts
+        .for_facility(product.facility)
+        .active.find_by_account_number(account_number)
+  end
+
+  def errors?
+    @errors.any?
+  end
+
+  def errors
+    @errors.to_a
+  end
+
+  def import
+    add_product_to_order if has_valid_fields?
+  end
+
+  def order_date
+    @order_date ||= parse_usa_import_date(order_date_field)
+  end
+
+  def order_id
+    order.id
+  end
+
+  def order_key
+    @order_key ||= [user_field, chart_string_field, order_date_field]
+  end
+
+  def row_with_errors # TODO refactor
+    new_row = @row.dup
+    new_row[HEADERS[:errors]] = errors.join(", ")
+    new_row
+  end
+
+  def user
+    @user ||=
+      User.find_by_username(user_field) || User.find_by_email(user_field)
+  end
+
+  private
+
+  def account_number
+    @account_number ||= @row[HEADERS[:chart_string]].strip
+  end
+
+  def add_error(message)
+    @errors.add(message) if message.present?
+  end
+
+  def add_product_to_order
+    @order_details = order.add(product, quantity, note: note)
+    purchase_order! unless order.purchased?
+    backdate_order_details_to_complete!
+  end
+
+  def backdate_order_details_to_complete!
+    @order_details.each do |order_detail|
+      order_detail.backdate_to_complete!(fulfillment_date)
+    end
+  end
+
+  def chart_string_field
+    @chart_string_field ||= @row[HEADERS[:chart_string]].strip
+  end
+
+  def fulfillment_date
+    @fulfillment_date ||= parse_usa_import_date(fulfillment_date_field)
+  end
+
+  def fulfillment_date_field
+    @fulfillment_date_field ||= @row[HEADERS[:fulfillment_date]].strip
+  end
+
+  def has_valid_fields?
+    validate_fields
+    !errors?
+  end
+
+  def note
+    @note ||= @row[HEADERS[:notes]].try(:strip)
+  end
+
+  def order
+    @order ||= @order_import.fetch_or_create_order!(self)
+  end
+
+  def order_date_field
+    @order_date_field ||= @row[HEADERS[:order_date]].strip
+  end
+
+  def product
+    @product ||=
+      @order_import
+        .facility
+        .products
+        .active_plus_hidden
+        .find_by_name(product_field)
+  end
+
+  def product_field
+    @product_field ||= @row[HEADERS[:product_name]].strip
+  end
+
+  def purchase_order!
+    if order.validate_order!
+      add_error("Couldn't purchase order") unless order.purchase!
+    else
+      add_error("Couldn't validate order")
+    end
+  end
+
+  def quantity
+    @quantity ||= @row[HEADERS[:quantity]].to_i
+  end
+
+  def user_field
+    @user_field ||= @row[HEADERS[:user]].strip
+  end
+
+  def validate_account
+    return if user.blank? || product.blank?
+    if account.present?
+      add_error(account.validate_against_product(product, user))
+    else
+      add_error("Can't find account")
+    end
+  end
+
+  def validate_fields
```

Comment:
Probably not worth it at this point, but we could probably use `ActiveModel::Validations` for this.